packages/git/src/browser/dirty-diff/dirty-diff-manager.ts

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable, postConstruct } from 'inversify';
+import { EditorManager, EditorWidget } from '@theia/editor/lib/browser';
+import { Workspace } from '@theia/languages/lib/common';
+import URI from '@theia/core/lib/common/uri';
+import { DirtyDiffComputer, DirtyDiff } from './diff-computer';
+import { Emitter, Event, Disposable, ReferenceCollection } from '@theia/core';
+import { DirtyDiffPreferences, DirtyDiffConfiguration } from './dirty-diff-preferences';
+import { PreferenceChangeEvent } from '@theia/core/lib/browser';
+import { GitResourceResolver, GIT_RESOURCE_SCHEME } from '../git-resource';
+import { WorkingDirectoryStatus, GitFileStatus, GitFileChange } from '../../common';
+import { GitRepositoryTracker } from '../git-repository-tracker';
+
+@injectable()
+export class DirtyDiffManager {
+
+    protected readonly models = new ReferenceCollection<string, DirtyDiffModel>(
+        uri => this.createNewModel(uri)
+    );
+
+    protected readonly onDityDiffUpdateEmitter = new Emitter<DityDiffUpdate>();
+    readonly onDirtyDiffUpdate: Event<DityDiffUpdate> = this.onDityDiffUpdateEmitter.event;
+
+    @inject(GitRepositoryTracker) protected readonly repositoryTracker: GitRepositoryTracker;
+    @inject(GitResourceResolver) protected readonly gitResourceResolver: GitResourceResolver;
+
+    @inject(EditorManager) protected readonly editorManager: EditorManager;
+
+    @inject(Workspace) protected readonly workspace: Workspace;
+    @inject(DirtyDiffPreferences) protected readonly preferences: DirtyDiffPreferences;
+
+    @postConstruct()
+    protected async initialize() {
+        this.workspace.onDidChangeTextDocument(async params => await this.handleDocumentChanged(params.textDocument.uri));
+        this.preferences.onPreferenceChanged(async e => await this.handlePreferenceChange(e));
+        this.editorManager.onCreated(async e => await this.handleEditorCreated(e));
+        this.repositoryTracker.onGitEvent(async event => await this.handleGitStatusUpdate(event.status));
+        const gitStatus = this.repositoryTracker.selectedRepositoryStatus;
+        if (gitStatus) {
+            await this.handleGitStatusUpdate(gitStatus);
+        }
+    }
+
+    protected async handleEditorCreated(editorWidget: EditorWidget): Promise<void> {
+        const editor = editorWidget.editor;
+        const uri = editor.document.uri;
+        const reference = await this.models.acquire(uri);
```

Comment:
We need to check, whether the file is actually under version control.

Reply:
Thanks @svenefftinge, this went broken with an optimization introduced recently. 🙄 it will show dirty diffs for files outside git repos. the [optimization](https://github.com/theia-ide/theia/blob/9d6ae3a93fa2a453f7aac436f6b49bec9e2b166d/packages/git/src/browser/dirty-diff/dirty-diff-manager.ts#L177-L179) targets the case of unmodified files to reduce delay of receiving git change events. but this is bogus, and I figured, how to change that.

Reply:
done.

model's previous content for unmodified files will be updated from git resource. the hacky solution using document's content is gone.